hedera-node/src/main/java/com/hedera/services/contracts/execution/EvmTxProcessor.java

```diff
@@ -323,41 +278,13 @@ protected TransactionProcessingResult execute(
 		}
 
 		// Externalise result
-		if (initialFrame.getState() == MessageFrame.State.COMPLETED_SUCCESS) {
-			return TransactionProcessingResult.successful(
-					initialFrame.getLogs(),
-					gasUsedByTransaction,
-					sbhRefund,
-					gasPrice,
-					initialFrame.getOutputData(),
-					mirrorReceiver,
-					stateChanges);
-		} else {
-			return TransactionProcessingResult.failed(
-					gasUsedByTransaction,
-					sbhRefund,
-					gasPrice,
-					initialFrame.getRevertReason(),
-					initialFrame.getExceptionalHaltReason(),
-					stateChanges);
-		}
-	}
-
-	private long calculateGasUsedByTX(final long txGasLimit, final MessageFrame initialFrame) {
-		long gasUsedByTransaction = txGasLimit - initialFrame.getRemainingGas();
-		/* Return leftover gas */
-		final long selfDestructRefund =
-				gasCalculator.getSelfDestructRefundAmount() *
-				Math.min(
-						initialFrame.getSelfDestructs().size(),
-						gasUsedByTransaction / (gasCalculator.getMaxRefundQuotient()));
-
-		gasUsedByTransaction = gasUsedByTransaction - selfDestructRefund - initialFrame.getGasRefund();
-
-		final var maxRefundPercent = dynamicProperties.maxGasRefundPercentage();
-		gasUsedByTransaction = Math.max(gasUsedByTransaction, txGasLimit - txGasLimit * maxRefundPercent / 100);
-
-		return gasUsedByTransaction;
+		final var isCompletedSuccess = initialFrame.getState() == MessageFrame.State.COMPLETED_SUCCESS;
+		return TxProcessorUtil.buildTransactionResult(
+				isCompletedSuccess, initialFrame.getLogs(),
+				updater.getSbhRefund(), gasUsedByTransaction,
+				gasPrice, mirrorReceiver, initialFrame.getOutputData(),
+				initialFrame.getRevertReason(), initialFrame.getExceptionalHaltReason(),
+				stateChanges);
```

Comment:
You can minimise the number of the arguments a lot if you do the following:

- Pass the initialFrame/messageFrame itself and then internally get the needed fields
- Instead of passing `isCompletedSuccess` you can also determine its value internally inside `buildTransactionResult`

Reply:
Unfortunately, MessageFrame and PrecompileMessage classes don't share a common parent, and can't be passed to the same method as arguments.